Improvements / Fixes for RadioGroup and apply to more windows

yw4z opened the pull request:
Tested on windows 11

### RadioGroup
• Added Disable() / Enable() functionality. Its used on save preset dialog
• Added ability to SetTooltip() per each radio control
• Removed some of definations from loop to improve draw speed on Refresh()
• Adding disabling function lead to new problem. disabling wxPanel caused unwanted effect on wxStaticText so i removed usage of it and switched to wxPanel / wxWindow and draw text with dc. But i got problems with disabling focus on child elements. Using wxBannerWindow fix that
•  Switched wxGenericStaticBitmap for icons. that fixes flickering problems. same functionality but has better buffering
before - video that shows flickering on controls
https://github.com/user-attachments/assets/94656a5b-e8ff-4870-af5f-7514e8300617
after - no flicker
https://github.com/user-attachments/assets/46ea709c-1419-4fa6-8ffc-3c08a0d9c434

• Limited focus rectangle with text frame. this will match focus style with checkboxes
before
![Screenshot-20250618105458](https://github.com/user-attachments/assets/a452e824-a7cc-40bf-b2dd-bc39786c284d)
after
![Screenshot-20250618105504](https://github.com/user-attachments/assets/e0ba4823-ab8c-4079-bf20-a7a5edc60498)
• Looks like using wxBannerWindow  and drawing focus rectangle is best solution if there are no problems with other platforms. I will also use same solution on improved checkbox class https://github.com/SoftFever/OrcaSlicer/pull/9907

### Save Preset Dialog
• Matched fonts
• Used radio buttons with vertical layout
Before
![Screenshot-20250618010611](https://github.com/user-attachments/assets/818e554a-4c2e-49fd-9792-8ddf5687052d)
After
![Screenshot-20250618010623](https://github.com/user-attachments/assets/2c1505e5-0407-4cb9-9351-90e7fa131ac7)
while disabled
![Screenshot-20250618105832](https://github.com/user-attachments/assets/2956cb50-e6a5-49ef-81d1-eeb166252c3a)

### Drop Project Dialog
• Removed background and matched fonts
Before
![Screenshot-20250618004156](https://github.com/user-attachments/assets/29bf48a2-af6d-4dff-b68a-945df0034bea)
After
![Screenshot-20250618003732](https://github.com/user-attachments/assets/dcafacc1-cd38-41d9-953e-839ebbcf01c9)

### Other Dialogs with RadioButtons
• All of them drawn by CreatePresetsDialog.cpp. and radio buttons drawn by same function. i didnt touched them in this PR since it requires more change and test
![Screenshot-20250618011339](https://github.com/user-attachments/assets/39d03254-bb04-44ec-a397-df6c20cbae4d)
![Screenshot-20250618110304](https://github.com/user-attachments/assets/0b941618-7db9-430c-bb31-cfd3516da040)
![Screenshot-20250618011330](https://github.com/user-attachments/assets/cecd10c1-68bc-4cc6-9891-dda059ff09e2)

yw4z commented:
@Noisyfox fixed height range on object list disappeared https://github.com/SoftFever/OrcaSlicer/issues/9942 in this commit. child items was using scroll window as parent. changed its parent to static box. let me know if there is better way

Noisyfox commented on src/slic3r/GUI/GUI_ObjectLayers.cpp:
```diff
@@ -27,7 +28,11 @@ ObjectLayers::ObjectLayers(wxWindow* parent) :
 
-    m_og->activate();
+    //m_og->activate();
+    LabeledStaticBox* stb = new LabeledStaticBox(parent);
```
Not sure if this is needed, since this is exactly what `m_og->activate()` does except the `SetCornerRadius`.

Noisyfox commented on src/slic3r/GUI/GUI_ObjectLayers.cpp:
```diff
@@ -74,7 +79,7 @@ wxSizer* ObjectLayers::create_layer(const t_layer_height_range& range, PlusMinus
 
     // Add text
-    auto head_text = new wxStaticText(m_parent, wxID_ANY, _L("Height Range"), wxDefaultPosition, wxDefaultSize, wxST_ELLIPSIZE_END);
+    auto head_text = new wxStaticText(m_og->stb, wxID_ANY, _L("Height Range"), wxDefaultPosition, wxDefaultSize, wxST_ELLIPSIZE_END);
```
I think we should use `m_og->ctrl_parent()` instead of use `stb` directly.

Noisyfox commented:
> Adding Hide and Show fix that. should i apply that in this PR

What if you use freeze & thaw?

yw4z commented:
yes that one proper method. sent a commit now

Noisyfox commented on src/slic3r/GUI/GUI_ObjectLayers.cpp:
```diff
@@ -30,6 +32,8 @@ ObjectLayers::ObjectLayers(wxWindow* parent) :
     m_og->activate();
     m_og->sizer->Clear(true);
     m_og->sizer->Add(m_grid_sizer, 0, wxEXPAND | wxLEFT | wxRIGHT, wxOSX ? 0 : 5);
```
`wxOSX ?` need to be removed too. Similar to previous PR.

Noisyfox replied:
Odd... Still not working properly even with this removed. Hold on.

Noisyfox commented on src/slic3r/GUI/Widgets/LabeledStaticBox.cpp:
```diff
@@ -63,11 +57,10 @@ bool LabeledStaticBox::Create(
     int tW,tH,descent,externalLeading;
-    GetTextExtent("Yy", &tW, &tH, &descent, &externalLeading, &m_font);
+    // empty label sets m_label_height as 0 that causes extra spacing at top
+    GetTextExtent(m_label.IsEmpty() ? "Orca" : m_label, &tW, &tH, &descent, &externalLeading, &m_font);
```
Actually it's this part that breaks the marging on macOS. I guess somehow it only removes the spacing on top but not bottom.

yw4z replied:
actually top part is looks same as windows but it adds extra margin to bottom
both OS starts client area on same position if it has title or not. so spacing on top expected

yw4z replied:
actually looks like macOS has extra margin inside static box, so client area starts from 0,0 and it has extra spacing on all edges. can you try this solution previously suggested https://github.com/SoftFever/OrcaSlicer/pull/9797#issuecomment-2956904196
i guess your implementation on here https://github.com/SoftFever/OrcaSlicer/pull/9797#issuecomment-2959401163 just removed margin outside of staticbox not inside padding. i hope both solution combined will remove all margin and padding

yw4z replied:
you can see extra padding on macOS wit side by side comparison with windows
![Screenshot-20250621102132](https://github.com/user-attachments/assets/8d63c1b8-4678-473a-af4a-340f28a5e4a8)

Noisyfox replied:
I think we need to override the `GetBordersForSizer` function for macOS:

![image](https://github.com/user-attachments/assets/dd9893a0-647c-43ec-a32a-d649e4d44081)

Noisyfox replied:
![image](https://github.com/user-attachments/assets/8f7248b4-1e7e-4204-9af1-2dcbf28c57e2)

Noisyfox commented:
The radio box steals the focus on the save preset dialog on macOS:
![radio-steal-focus](https://github.com/user-attachments/assets/c0e74128-b760-4f81-a66d-a9853af07efa)

yw4z commented:
tested that problem on windows and its not exist. you build from this PR right because i saw a similar bug without this PR https://github.com/SoftFever/OrcaSlicer/pull/9809 applied. buttons was stealing focus from input box in that case
![orca-slicer_Me5JadbYNB](https://github.com/user-attachments/assets/88ebff0e-7635-47bf-9d54-c8599c6dbd43)

yw4z commented:
sent a new commit for possible fix. maybe focus not resets on refresh

replaced `if (m_focused)` with `if (m_focused && !HasFocus())` . this should fix if it causes an infinite loop

yw4z commented:
i have tried that but while creating RadioGroup but ofc i didnt see an effect since there are no problems and only tested on windows.
should we call it on every OnClick()

Noisyfox commented:
> should we call it on every OnClick()

I haven't find any issue so far.

Noisyfox commented:
Unfortunately this doesn't work on Linux...

Noisyfox commented:
I think wxpanel just don't like been focused... Maybe consider add focus listener to the text & icons instead?

yw4z commented:
RadioGroup should use only 1 focus element. I have tried that approach before.
Enabled focus for only one of radio button then i disabled for others then enabled focus for clicked item and disabled for others.
problem appears when moving focus to other radio button
I have to disable focus for that element with DisableFocusFromKeyboard() then AcceptFocusFromKeyboard() for selected element that never worked properly or made code more complex

We can find problematic part since current / previous code works. here is changes;
• Main control same as before wxPanel. that worked previously on Linux and macOS
• removed ` if (m_focused)  SetFocus();` part from wxPanel's paint event. that worked previously but macOS stucks on focus loop with this changes
• Replaced wxStaticText with wxBannerWindow and moved draw of focus rectangle to its paint event. this one is required for both reasons; disabling wxStaticText has an unwanted effect and wxStaticText overlaps focus rectangle. wxBannerWindow  should not steal any focus according to documentation. It designed for Non-Interactivity. Maybe adding wxEVT_LEFT_DOWN with ` e.Skip();` cause issues since we dont know fully` e.Skip();` what it does for each platform
• Replaced wxStaticBitmap with wxGenericStaticBitmap. it fixes flickering and it should not create any problems with focus

did you tried re adding SetFocus() in paint of wxPanel with combined SetFocusIgnoringChildren()
```
 if (m_focused) 
    SetFocus();
``` 

maybe removing ` e.Skip();`  from wxGenericStaticBitmap and wxBannerWindow can make a difference

Alternative path we can add an dummy wxButton / wxTextCtrl inside. but i'm not sure we can set 1x1px size in all platforms to make it invincible

Noisyfox commented:
> The text field is working, but this issue persists: ![image](https://private-user-images.githubusercontent.com/1537155/457990302-496e4474-ad12-421d-a0cd-d98a8cf8432d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE4MDc0NjksIm5iZiI6MTc1MTgwNzE2OSwicGF0aCI6Ii8xNTM3MTU1LzQ1Nzk5MDMwMi00OTZlNDQ3NC1hZDEyLTQyMWQtYTBjZC1kOThhOGNmODQzMmQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDZUMTMwNjA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmNmZmE5MDg5YTRjMzQ4ZTY5MDQxZjFiZTg2NDhhYzAxNTFkNjljYWJmMjU5ZjExM2IxM2FiNDBjMzQ0MTllMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.iSs3z93hAJKO7T_nuX6NKiSoMCmA0b1nlTf7cVbDIT0)

This issue is present on Linux too and could not be fixed with anything we've mentioned above, expect by removing the changes to RadioGroup.

yw4z commented:
Sent new commit, added some minor changes that can be solution
- added SetFocusIgnoringChildren() while creation of wxPanel
- defined SetCanFocus() for each control. read a comment this can be helpful on macOS for wxPanel
- used wxAutoBufferedPaintDC  instead wxPaintDC .read a comment this can be helpful .wxPanel double buffered as default on macOS
can you check these 
- how many focusable controls on that dialog. RadioGroup should only take one. maybe problem is just dc.Clear() not working on control painting

few more things can be solution
- using dc.SetBackground(*wxWHITE_BRUSH); dc.DrawRectangle(GetClientRect()); if problems related with dc.clear()
- removing SetDoubleBuffered(true); or using it only for windows
- saw a macOS specific solution on CheckBox.cpp
https://github.com/SoftFever/OrcaSlicer/blob/b785f40fb036ac4a804155f1bed4acb29d1a4251/src/slic3r/GUI/Widgets/CheckBox.cpp#L21-L26
https://github.com/SoftFever/OrcaSlicer/blob/b785f40fb036ac4a804155f1bed4acb29d1a4251/src/slic3r/GUI/Widgets/CheckBox.cpp#L104-L123
- usage on TextInput.cpp but i assume this is not the case for wxPanel since children not taking focus
https://github.com/SoftFever/OrcaSlicer/blob/b785f40fb036ac4a804155f1bed4acb29d1a4251/src/slic3r/GUI/Widgets/TextInput.cpp#L67-L72

Noisyfox commented:
Close in favor of #10089
